// @mablhq/mabl-cli/commands/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandArgMablBranchChangesOnly = exports.CommandArgMablBranch = exports.CommandArgMablAutoLogin = exports.CommandArgMablAutoBranch = exports.CommandArgLinkBypass = exports.CommandArgLinkLabel = exports.CommandArgLimitOutput = exports.CommandArgLabels = exports.CommandArgLabelsInclude = exports.CommandArgLabelsExclude = exports.CommandArgTestInteractionSpeed = exports.CommandArgTestRunId = exports.CommandArgTestFile = exports.CommandArgKeepBrowserOpen = exports.CommandArgInput = exports.CommandArgIncludedTests = exports.CommandArgIncludeDefaults = exports.CommandArgId = exports.CommandArgUserAgent = exports.CommandArgHttpHeaders = exports.CommandArgHelp = exports.CommandArgHeadless = exports.CommandArgFromPlanId = exports.CommandArgFrom = exports.CommandArgFormat = exports.CommandArgFindPath = exports.CommandArgFastFailure = exports.CommandArgOverrideEnvironmentId = exports.CommandArgGrep = exports.CommandArgExtraArguments = exports.CommandArgExistingReport = exports.CommandArgExcludedTests = exports.CommandArgEnvironmentId = exports.CommandArgDetailLevel = exports.CommandArgDestination = exports.CommandArgDescription = exports.CommandArgDeploymentId = exports.CommandArgDecrypt = exports.CommandArgDebug = exports.CommandArgDataTables = exports.CommandArgDataTableId = exports.CommandArgCredentials = exports.CommandArgBasicAuthCredentials = exports.CommandArgBrowsers = exports.CommandArgBrowser = exports.CommandArgBranch = exports.CommandArgAuto = exports.CommandArgApplicationId = exports.CommandArgApiKey = exports.CommandArgABConfigFile = void 0;
exports.DefaultBrowserType = exports.DefaultBranchName = exports.DefaultOutputFormatChoices = exports.DetailLevelFormats = exports.Reporter = exports.OutputFormats = exports.CommandArgAliases = exports.CommandArgBrowserEnableExtensions = exports.CommandArgBrowserIgnoreCertificateErrors = exports.CommandArgBrowserDisableIsolation = exports.ListTimeFormat = exports.CommandArgTimezoneID = exports.CommandArgLocale = exports.CommandArgUseTestExecutionProxy = exports.CommandArgScenarioId = exports.CommandArgReporterOptions = exports.CommandArgReporter = exports.CommandArgWorkspaceId = exports.CommandArgVersion = exports.CommandArgVariables = exports.CommandArgUrlApi = exports.CommandArgUrlApp = exports.CommandArgUrl = exports.CommandArgTo = exports.CommandArgTrainerVersion = exports.CommandArgTraceFile = exports.CommandArgTracesPath = exports.CommandArgTestPath = exports.CommandArgSilent = exports.CommandArgRevision = exports.CommandArgRecordVideoPath = exports.CommandArgMaxHeartbeatAge = exports.CommandArgPlanId = exports.CommandArgContentTypes = exports.CommandArgPreview = exports.CommandArgPath = exports.CommandArgOutputFilePath = exports.CommandArgOutput = exports.CommandArgNoPrompt = exports.CommandArgPrompt = exports.CommandArgProject = exports.CommandArgPortNumber = exports.CommandArgName = exports.CommandArgMobileDeviceName = exports.CommandArgMobilePlatform = exports.CommandArgMobileBuildFileId = exports.CommandArgMobileBuildFilePath = exports.CommandArgiOSdMobileAppFileId = exports.CommandArgAndroidMobileAppFileId = exports.CommandArgMulti = void 0;
exports.SCENARIO_ID_HEADER = exports.SCENARIO_NAME_HEADER = exports.ValidBrowserTypesForLocalRuns = exports.BrowserTypeSelections = void 0;
const browserTypes_1 = require("./browserTypes");
exports.CommandArgABConfigFile = 'ab-config-file';
exports.CommandArgApiKey = 'api-key';
exports.CommandArgApplicationId = 'application-id';
exports.CommandArgAuto = 'auto-save';
exports.CommandArgBranch = 'branch';
exports.CommandArgBrowser = 'browser';
exports.CommandArgBrowsers = 'browsers';
exports.CommandArgBasicAuthCredentials = 'basic-auth-credentials-id';
exports.CommandArgCredentials = 'credentials-id';
exports.CommandArgDataTableId = 'data-table-id';
exports.CommandArgDataTables = 'data-tables';
exports.CommandArgDebug = 'debug';
exports.CommandArgDecrypt = 'decrypt';
exports.CommandArgDeploymentId = 'deployment-id';
exports.CommandArgDescription = 'description';
exports.CommandArgDestination = 'destination';
exports.CommandArgDetailLevel = 'detail-level';
exports.CommandArgEnvironmentId = 'environment-id';
exports.CommandArgExcludedTests = 'excluded-tests';
exports.CommandArgExistingReport = 'existing-report';
exports.CommandArgExtraArguments = 'extra-arguments';
exports.CommandArgGrep = 'grep';
exports.CommandArgOverrideEnvironmentId = 'override-environment-id';
exports.CommandArgFastFailure = 'fast-failure';
exports.CommandArgFindPath = 'find-path';
exports.CommandArgFormat = 'format';
exports.CommandArgFrom = 'from';
exports.CommandArgFromPlanId = 'from-plan-id';
exports.CommandArgHeadless = 'headless';
exports.CommandArgHelp = 'help';
exports.CommandArgHttpHeaders = 'http-headers';
exports.CommandArgUserAgent = 'user-agent';
exports.CommandArgId = 'id';
exports.CommandArgIncludeDefaults = 'include-defaults';
exports.CommandArgIncludedTests = 'included-tests';
exports.CommandArgInput = 'input';
exports.CommandArgKeepBrowserOpen = 'keep-browser-open';
exports.CommandArgTestFile = 'test-file';
exports.CommandArgTestRunId = 'run-id';
exports.CommandArgTestInteractionSpeed = 'interaction-speed';
exports.CommandArgLabelsExclude = 'exclude-labels';
exports.CommandArgLabelsInclude = 'labels';
exports.CommandArgLabels = 'labels';
exports.CommandArgLimitOutput = 'limit';
exports.CommandArgLinkLabel = 'link';
exports.CommandArgLinkBypass = 'link-bypass';
exports.CommandArgMablAutoBranch = 'auto-branch';
exports.CommandArgMablAutoLogin = 'auto-login';
exports.CommandArgMablBranch = 'mabl-branch';
exports.CommandArgMablBranchChangesOnly = 'branch-changes-only';
exports.CommandArgMulti = 'multi';
exports.CommandArgAndroidMobileAppFileId = 'android-app-file-id';
exports.CommandArgiOSdMobileAppFileId = 'ios-app-file-id';
exports.CommandArgMobileBuildFilePath = 'build-file-path';
exports.CommandArgMobileBuildFileId = 'build-file-id';
exports.CommandArgMobilePlatform = 'platform';
exports.CommandArgMobileDeviceName = 'device';
exports.CommandArgName = 'name';
exports.CommandArgPortNumber = 'port';
exports.CommandArgProject = 'project';
exports.CommandArgPrompt = 'prompt';
exports.CommandArgNoPrompt = `no-${exports.CommandArgPrompt}`;
exports.CommandArgOutput = 'output';
exports.CommandArgOutputFilePath = 'file';
exports.CommandArgPath = 'path';
exports.CommandArgPreview = 'preview';
exports.CommandArgContentTypes = 'types';
exports.CommandArgPlanId = 'plan-id';
exports.CommandArgMaxHeartbeatAge = 'max-heartbeat-age';
exports.CommandArgRecordVideoPath = 'video-path';
exports.CommandArgRevision = 'revision';
exports.CommandArgSilent = 'silent';
exports.CommandArgTestPath = 'tests-path';
exports.CommandArgTracesPath = 'traces-path';
exports.CommandArgTraceFile = 'trace-file';
exports.CommandArgTrainerVersion = 'trainer-version';
exports.CommandArgTo = 'to';
exports.CommandArgUrl = 'url';
exports.CommandArgUrlApp = 'app-url';
exports.CommandArgUrlApi = 'api-url';
exports.CommandArgVariables = 'variables';
exports.CommandArgVersion = 'version';
exports.CommandArgWorkspaceId = 'workspace-id';
exports.CommandArgReporter = 'reporter';
exports.CommandArgReporterOptions = 'reporter-options';
exports.CommandArgScenarioId = 'scenario-id';
exports.CommandArgUseTestExecutionProxy = 'use-test-execution-proxy';
exports.CommandArgLocale = 'locale';
exports.CommandArgTimezoneID = 'timezone-id';
exports.ListTimeFormat = 'MMM Do YYYY, HH:mm:ss';
exports.CommandArgBrowserDisableIsolation = 'browser-disable-isolation';
exports.CommandArgBrowserIgnoreCertificateErrors = 'browser-ignore-certificate-errors';
exports.CommandArgBrowserEnableExtensions = 'enable-browser-extensions';
var CommandArgAliases;
(function (CommandArgAliases) {
    CommandArgAliases["ApiKey"] = "k";
    CommandArgAliases["ApplicationId"] = "a";
    CommandArgAliases["BasicAuthCredentials"] = "basic-creds";
    CommandArgAliases["BranchFrom"] = "f";
    CommandArgAliases["BranchTo"] = "t";
    CommandArgAliases["Browsers"] = "b";
    CommandArgAliases["Credentials"] = "creds";
    CommandArgAliases["DeploymentId"] = "d";
    CommandArgAliases["EnvironmentId"] = "e";
    CommandArgAliases["File"] = "f";
    CommandArgAliases["LimitOutput"] = "l";
    CommandArgAliases["Name"] = "n";
    CommandArgAliases["NoPrompt"] = "yes";
    CommandArgAliases["OutputType"] = "o";
    CommandArgAliases["Revision"] = "rev";
    CommandArgAliases["ShowHelp"] = "h";
    CommandArgAliases["Status"] = "s";
    CommandArgAliases["Url"] = "u";
    CommandArgAliases["Version"] = "v";
    CommandArgAliases["WorkspaceId"] = "w";
})(CommandArgAliases || (exports.CommandArgAliases = CommandArgAliases = {}));
var OutputFormats;
(function (OutputFormats) {
    OutputFormats["Csv"] = "csv";
    OutputFormats["Json"] = "json";
    OutputFormats["Markdown"] = "markdown";
    OutputFormats["Playwright"] = "playwright";
    OutputFormats["SeleniumIde"] = "side";
    OutputFormats["Table"] = "table";
    OutputFormats["Yaml"] = "yaml";
})(OutputFormats || (exports.OutputFormats = OutputFormats = {}));
var Reporter;
(function (Reporter) {
    Reporter["Mabl"] = "mabl";
    Reporter["Mochawesome"] = "mochawesome";
})(Reporter || (exports.Reporter = Reporter = {}));
var DetailLevelFormats;
(function (DetailLevelFormats) {
    DetailLevelFormats["Full"] = "full";
})(DetailLevelFormats || (exports.DetailLevelFormats = DetailLevelFormats = {}));
exports.DefaultOutputFormatChoices = [
    OutputFormats.Json,
    OutputFormats.Yaml,
];
exports.DefaultBranchName = 'master';
exports.DefaultBrowserType = browserTypes_1.BrowserType.Chrome;
exports.BrowserTypeSelections = [
    browserTypes_1.BrowserType.Chrome,
    browserTypes_1.BrowserType.Edge,
    browserTypes_1.BrowserType.Firefox,
    browserTypes_1.BrowserType.Webkit,
].map((browser) => browser.toString());
exports.ValidBrowserTypesForLocalRuns = [
    browserTypes_1.BrowserType.Chrome,
    browserTypes_1.BrowserType.Edge,
    browserTypes_1.BrowserType.Firefox,
    browserTypes_1.BrowserType.Webkit,
];
exports.SCENARIO_NAME_HEADER = 'Scenario name';
exports.SCENARIO_ID_HEADER = 'Scenario ID';
