// @mablhq/mabl-cli/core/execution/newman-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BodyMode = exports.MablSupportedPostmanAuthTypes = exports.SUPPORTED_COLLECTION_SCHEMA = exports.AssertionType = exports.AssertionTarget = void 0;
var AssertionTarget;
(function (AssertionTarget) {
    AssertionTarget["Header"] = "Header";
    AssertionTarget["JSONBody"] = "JSONBody";
    AssertionTarget["Size"] = "Size";
    AssertionTarget["Status"] = "Status";
    AssertionTarget["TextBody"] = "TextBody";
})(AssertionTarget || (exports.AssertionTarget = AssertionTarget = {}));
var AssertionType;
(function (AssertionType) {
    AssertionType["Contains"] = "Contains";
    AssertionType["DoesNotContain"] = "DoesNotContain";
    AssertionType["EndsWith"] = "EndsWith";
    AssertionType["Equals"] = "Equals";
    AssertionType["GreaterThan"] = "GreaterThan";
    AssertionType["GreaterThanOrEqualTo"] = "GreaterThanOrEqualTo";
    AssertionType["LessThan"] = "LessThan";
    AssertionType["LessThanOrEqualTo"] = "LessThanOrEqualTo";
    AssertionType["MatchesRegExp"] = "MatchesRegExp";
    AssertionType["NotEquals"] = "NotEquals";
    AssertionType["NotPresent"] = "NotPresent";
    AssertionType["Present"] = "Present";
    AssertionType["StartsWith"] = "StartsWith";
})(AssertionType || (exports.AssertionType = AssertionType = {}));
exports.SUPPORTED_COLLECTION_SCHEMA = 'https://schema.getpostman.com/json/collection/v2.1.0/collection.json';
exports.MablSupportedPostmanAuthTypes = [
    'apikey',
    'basic',
    'bearer',
    'oauth1',
];
var BodyMode;
(function (BodyMode) {
    BodyMode["file"] = "file";
    BodyMode["formdata"] = "formdata";
    BodyMode["graphql"] = "graphql";
    BodyMode["none"] = "none";
    BodyMode["raw"] = "raw";
    BodyMode["urlencoded"] = "urlencoded";
})(BodyMode || (exports.BodyMode = BodyMode = {}));
