// @mablhq/mabl-cli/mablscript/types/mobile/ScrollStepDescriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isScrollDeviceStepDescriptor = exports.isScrollWithinElementStepDescriptor = exports.isScrollToElementStepDescriptor = exports.ScrollType = exports.ScrollDirection = void 0;
var ScrollDirection;
(function (ScrollDirection) {
    ScrollDirection["Down"] = "down";
    ScrollDirection["Left"] = "left";
    ScrollDirection["Right"] = "right";
    ScrollDirection["Up"] = "up";
})(ScrollDirection || (exports.ScrollDirection = ScrollDirection = {}));
var ScrollType;
(function (ScrollType) {
    ScrollType["ScrollDeviceScreen"] = "scroll_device_screen";
    ScrollType["ScrollToElement"] = "scroll_to_element";
    ScrollType["ScrollWithinElement"] = "scroll_within_element";
})(ScrollType || (exports.ScrollType = ScrollType = {}));
function isScrollToElementStepDescriptor(step) {
    return (step.scrollType === ScrollType.ScrollToElement &&
        !!step.scrollDirection &&
        !!step.targetElementDescriptor);
}
exports.isScrollToElementStepDescriptor = isScrollToElementStepDescriptor;
function isScrollWithinElementStepDescriptor(step) {
    return (step.scrollType === ScrollType.ScrollWithinElement &&
        !!step.scrollDirection &&
        !!step.find);
}
exports.isScrollWithinElementStepDescriptor = isScrollWithinElementStepDescriptor;
function isScrollDeviceStepDescriptor(step) {
    return (step.scrollType === ScrollType.ScrollDeviceScreen && !!step.scrollDirection);
}
exports.isScrollDeviceStepDescriptor = isScrollDeviceStepDescriptor;
