// @mablhq/mabl-cli/mablscript/steps/ClickAndHoldStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickAndHoldStep = void 0;
const FindAction_1 = require("../actions/FindAction");
const domUtil_1 = require("../../domUtil");
const ActionsUtils_1 = require("./ActionsUtils");
const MablStepWithFindAction_1 = require("../MablStepWithFindAction");
class ClickAndHoldStep extends MablStepWithFindAction_1.MablStepWithFindAction {
    constructor(name, args, actions) {
        var _a;
        super(name, args, actions, 'click_and_hold');
        this.findAction = ActionsUtils_1.ActionsUtils.validateSingleFindAction(this.actions);
        this.clickArgs = (_a = this.parseArgs(args)) !== null && _a !== void 0 ? _a : { isHtml5: false };
    }
    parseArgs(args) {
        if (args.length === 0) {
            return undefined;
        }
        if (args.length !== 1) {
            throw new Error(`Unexpected number of arguments to ${this.getStepName()} step.`);
        }
        if (!isClickAndHoldArgs(args[0])) {
            throw new Error(`Invalid argument passed to ${this.getStepName()} step.`);
        }
        return args[0];
    }
    getStepName() {
        return 'ClickAndHold';
    }
    toStepDescriptor() {
        const find = this.findAction.toDescriptor();
        switch (find.findType) {
            case domUtil_1.FindType.FIND_FIRST:
            case domUtil_1.FindType.FIND_LAST:
            case domUtil_1.FindType.FIND_ANY:
            case domUtil_1.FindType.FIND_ONE:
                return {
                    find,
                    isHtml5: this.clickArgs.isHtml5,
                    descriptorToActionMap: new Map().set(find, this.findAction),
                    actionCode: this.actionCode,
                };
            default:
                throw new Error(`Error generating step descriptor for ${this.getStepName()}: Unexpected find type ${find.findType}`);
        }
    }
    getFormattedStep(_fullLocatorsOn) {
        const step = {
            ClickAndHold: {
                ...super.annotationsAsYml(),
                ...this.findAction.toYaml(),
            },
        };
        if (this.stepId()) {
            step.ClickAndHold.id = this.stepId();
        }
        return step;
    }
    static fromYaml(_stepName, stepArgs) {
        const step = new ClickAndHoldStep('click_and_hold', [], [FindAction_1.FindAction.findActionFromStepArgs(stepArgs)]);
        step.setStepId(stepArgs.id);
        return step;
    }
    toMablscript() {
        return `${this.findAction.toMablscript()}.click_and_hold({isHtml5:${this.clickArgs.isHtml5}})`;
    }
    getInputVariables() {
        return this.findAction.getInputVariables();
    }
}
exports.ClickAndHoldStep = ClickAndHoldStep;
ClickAndHoldStep.mablScriptStepNames = ['click_and_hold'];
ClickAndHoldStep.yamlMablScriptNames = ['ClickAndHold'];
function isClickAndHoldArgs(value) {
    return value.isHtml5 !== undefined;
}
