// @mablhq/mabl-cli/mablscript/steps/EchoStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EchoStep = void 0;
const MablStep_1 = require("../MablStep");
const domUtil_1 = require("../../domUtil");
const MablAction_1 = require("../MablAction");
class EchoStep extends MablStep_1.MablStep {
    constructor(name, args, actions) {
        super(name, args, actions, 'echo');
        this.echoValue = (0, MablAction_1.parseArgument)(this.getActionArgs()[0]);
    }
    getStepName() {
        return EchoStep.stepName;
    }
    toStepDescriptor() {
        return { value: this.echoValue.toString(), actionCode: this.actionCode };
    }
    static fromYaml(_stepName, stepArgs) {
        const step = new EchoStep(EchoStep.stepName, [stepArgs.value], []);
        step.setStepId(stepArgs.id);
        return step;
    }
    toMablscript() {
        const stringEchoValue = `${this.echoValue}`;
        return `echo("${(0, domUtil_1.escapeMablscriptString)(stringEchoValue)}")`;
    }
    getInputVariables() {
        return MablAction_1.MablAction.findUniqueVariableReferencesInValue(this.echoValue);
    }
}
exports.EchoStep = EchoStep;
EchoStep.stepName = 'Echo';
EchoStep.mablScriptStepNames = ['echo'];
EchoStep.yamlMablScriptNames = [EchoStep.stepName];
