// @mablhq/mabl-cli/mablscript/steps/RemoveCookieStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveCookieStep = void 0;
const MablStep_1 = require("../MablStep");
const MablAction_1 = require("../MablAction");
const CookieUtils_1 = require("./CookieUtils");
const domUtil_1 = require("../../domUtil");
class RemoveCookieStep extends MablStep_1.MablStep {
    constructor(name, args, actions) {
        super(name, args, actions, 'remove_cookie');
        this.cookieName = CookieUtils_1.CookieUtils.normalizeName(this.getActionArgs()[0].name);
    }
    getStepName() {
        return 'RemoveCookie';
    }
    toStepDescriptor() {
        return {
            name: this.cookieName,
            actionCode: this.actionCode,
        };
    }
    static fromYaml(_stepName, stepArgs) {
        const step = new RemoveCookieStep('remove_cookie', [stepArgs], []);
        step.setStepId(stepArgs.id);
        return step;
    }
    toMablscript() {
        return `remove_cookie({ name : "${(0, domUtil_1.escapeMablscriptString)(this.cookieName)}" })`;
    }
    getInputVariables() {
        return MablAction_1.MablAction.findUniqueVariableReferencesInValue(this.cookieName);
    }
}
exports.RemoveCookieStep = RemoveCookieStep;
RemoveCookieStep.mablScriptStepNames = ['remove_cookie'];
RemoveCookieStep.yamlMablScriptNames = ['RemoveCookie'];
