// @mablhq/mabl-cli/mablscript/steps/SetViewportStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetViewportStep = void 0;
const MablStep_1 = require("../MablStep");
class SetViewportStep extends MablStep_1.MablStep {
    constructor(name, args, actions) {
        super(name, args, actions, 'set_viewport');
        const viewportArg = this.getActionArgs()[0];
        this.size = viewportArg === null || viewportArg === void 0 ? void 0 : viewportArg.size;
    }
    getStepName() {
        return 'SetViewport';
    }
    toStepDescriptor() {
        return { size: this.size, actionCode: this.actionCode };
    }
    static fromYaml(_stepName, stepArgs) {
        const step = new SetViewportStep('set_viewport', [stepArgs], []);
        step.setStepId(stepArgs.id);
        return step;
    }
    toMablscript() {
        let sizeArg;
        if (this.size.height && this.size.width) {
            sizeArg = `{ size : { width : ${this.size.width}, height : ${this.size.height} } }`;
        }
        else if (this.size.width) {
            sizeArg = `{ size : { width : ${this.size.width} } }`;
        }
        else if (this.size.height) {
            sizeArg = `{ size : { height : ${this.size.height} } }`;
        }
        return `set_viewport(${sizeArg})`;
    }
}
exports.SetViewportStep = SetViewportStep;
SetViewportStep.mablScriptStepNames = ['set_viewport'];
SetViewportStep.yamlMablScriptNames = ['SetViewport'];
