// @mablhq/mabl-cli/env/local.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FIND_OVERRIDE_PATH = exports.ELECTRON_PROTOCOL = exports.SCRIPT_NAME = exports.OKTA_URL = exports.OKTA_CLIENT_ID = exports.LOCAL_TRAINER_PATH = exports.EXTENSION_ID = exports.ENV = exports.DEBUG_EVENT_EMITTER = exports.CONSOLE_LOGGING_LEVEL = exports.CONF_FILE_VERSION = exports.CONF_FILE_PROJECT_NAME = exports.BASE_APP_URL = exports.BASE_API_URL = void 0;
exports.BASE_API_URL = 'http://localhost:8080';
exports.BASE_APP_URL = 'https://app-local.mabl.com:3000';
exports.CONF_FILE_PROJECT_NAME = 'mabl-cli-local';
exports.CONF_FILE_VERSION = '1';
exports.CONSOLE_LOGGING_LEVEL = 'debug';
exports.DEBUG_EVENT_EMITTER = false;
exports.ENV = 'local';
exports.EXTENSION_ID = 'hlnobmannnbojpffhcdiljnmkfjpkkfn';
exports.LOCAL_TRAINER_PATH = process.env.LOCAL_TRAINER_PATH;
exports.OKTA_CLIENT_ID = '0oamuehnyLZhBbKh91d6';
exports.OKTA_URL = 'https://auth2-dev.mabl.com';
exports.SCRIPT_NAME = 'mabl-local';
exports.ELECTRON_PROTOCOL = 'mabl-app-local';
exports.FIND_OVERRIDE_PATH = undefined;
