// @mablhq/mabl-cli/commands/applications/applications_cmds/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const js_yaml_1 = require("js-yaml");
const mablApiClientFactory_1 = require("../../../api/mablApiClientFactory");
const util_1 = require("../../commandUtil/util");
const cli_table3_1 = __importDefault(require("cli-table3"));
const moment = require("moment");
const list_1 = require("../../commandUtil/list");
const loggingProvider_1 = require("../../../providers/logging/loggingProvider");
const constants_1 = require("../../constants");
exports.command = 'list';
exports.describe = 'List your applications';
exports.builder = (0, list_1.getListBuilderOptions)('applications');
exports.handler = (0, util_1.failWrapper)(listApplications);
async function listApplications(parsed) {
    const output = parsed.output;
    const limit = parsed.limit;
    const workspaceId = await (0, util_1.getWorkspaceId)(parsed);
    const apiClient = await mablApiClientFactory_1.MablApiClientFactory.createApiClient();
    const applications = await apiClient.getApplications(workspaceId, limit);
    printApplications(applications, output);
    return applications.length;
}
function printApplications(applications, output) {
    switch (output) {
        case 'json':
            loggingProvider_1.logger.info(JSON.stringify(applications, null, 2));
            break;
        case 'yaml':
            loggingProvider_1.logger.info((0, js_yaml_1.dump)(applications));
            break;
        default:
            const table = new cli_table3_1.default({
                head: ['ID', 'Name', 'Created time'],
                wordWrap: true,
            });
            applications.forEach((application) => {
                table.push([
                    { rowSpan: 1, content: application.id, vAlign: 'center' },
                    { rowSpan: 1, content: application.name, vAlign: 'center' },
                    {
                        rowSpan: 1,
                        content: moment
                            .utc(application.created_time)
                            .format(constants_1.ListTimeFormat),
                        vAlign: 'center',
                    },
                ]);
            });
            loggingProvider_1.logger.info(table.toString());
            break;
    }
}
