// @mablhq/mabl-cli/util/InternalMetricsTrackingSingleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalMetricsTrackingSingleton = void 0;
const logUtils_1 = require("./logUtils");
class InternalMetricsTrackingSingleton {
    constructor() {
        this.metadata = {};
    }
    static getInstance() {
        if (!InternalMetricsTrackingSingleton.instance) {
            InternalMetricsTrackingSingleton.instance =
                new InternalMetricsTrackingSingleton();
        }
        return InternalMetricsTrackingSingleton.instance;
    }
    addMetric(name, value) {
        this.metadata[name] = value;
    }
    addMetrics(metrics) {
        this.metadata = { ...this.metadata, ...metrics };
    }
    testType(testType) {
        this.metadata.testType = testType;
    }
    testStatus(testStatus) {
        this.metadata.testStatus = testStatus;
    }
    reportMetrics() {
        if (Object.keys(this.metadata).length === 0) {
            return;
        }
        (0, logUtils_1.logInternal)('Internal metrics:', this.metadata);
        this.metadata = {};
    }
}
exports.InternalMetricsTrackingSingleton = InternalMetricsTrackingSingleton;
