// @mablhq/mabl-cli/browserEngines/unsupportedBrowserEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsupportedBrowserEngine = void 0;
const mobileEmulationUtil_1 = require("../commands/tests/mobileEmulationUtil");
class UnsupportedBrowserEngine {
    name() {
        throw new Error('Unsupported browser engine');
    }
    getBrowserLaunchOptions(options) {
        return {
            commandLineArgs: [],
            ignoreCommandLineDefaultArgs: [],
            defaultDeviceDescriptor: (0, mobileEmulationUtil_1.getDeviceDescriptorForEmulation)(options.emulationConfig),
        };
    }
    getExecutionEngineBrowserLaunchOptions(_options, _proxyInfo) {
        throw new Error('Unsupported browser engine');
    }
    prepareBrowserPreferencesDirectory(_windowPlacement) {
        throw new Error('Unsupported browser engine');
    }
    findBrowserExecutable() {
        throw new Error('Unsupported browser engine');
    }
    getProxySpec(_proxyInfo) {
        throw new Error('Unsupported browser engine');
    }
}
exports.UnsupportedBrowserEngine = UnsupportedBrowserEngine;
