// @mablhq/mabl-cli/commands/plans/plans_cmds/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../../constants");
const util_1 = require("../../commandUtil/util");
const mablApiClientFactory_1 = require("../../../api/mablApiClientFactory");
const cli_table3_1 = __importDefault(require("cli-table3"));
const moment = require("moment");
const loggingProvider_1 = require("../../../providers/logging/loggingProvider");
const js_yaml_1 = require("js-yaml");
exports.command = 'list';
exports.describe = 'List plans';
exports.builder = (yargs) => {
    yargs
        .option(constants_1.CommandArgWorkspaceId, {
        alias: constants_1.CommandArgAliases.WorkspaceId,
        describe: 'Workspace to list plans for',
        nargs: 1,
        type: 'string',
    })
        .option(constants_1.CommandArgOutput, {
        alias: constants_1.CommandArgAliases.OutputType,
        default: constants_1.OutputFormats.Table,
        describe: 'Specify result output format',
        nargs: 1,
        choices: [...constants_1.DefaultOutputFormatChoices, constants_1.OutputFormats.Table],
    })
        .option(constants_1.CommandArgLimitOutput, {
        alias: constants_1.CommandArgAliases.LimitOutput,
        describe: 'The number of plans to return',
        default: 10,
        nargs: 1,
        type: 'string',
    });
};
exports.handler = (0, util_1.failWrapper)(listPlans);
async function listPlans(parsed) {
    const limit = parsed.limit;
    const workspaceId = await (0, util_1.getWorkspaceId)(parsed);
    const output = parsed[constants_1.CommandArgOutput];
    const apiClient = await mablApiClientFactory_1.MablApiClientFactory.createApiClient();
    const plans = await apiClient.getPlans({
        organization_id: workspaceId,
        limit,
    });
    const outputPlans = plans.map((plan) => ({
        created_time: plan.created_time,
        id: plan.id,
        name: plan.name,
    }));
    switch (output) {
        case constants_1.OutputFormats.Json:
            loggingProvider_1.logger.info(JSON.stringify(outputPlans, null, 2));
            break;
        case constants_1.OutputFormats.Yaml:
            loggingProvider_1.logger.info((0, js_yaml_1.dump)(outputPlans));
            break;
        case constants_1.OutputFormats.Table:
        default:
            printPlansAsTable(outputPlans);
            loggingProvider_1.logger.info(`${outputPlans.length} plans returned`);
            if (outputPlans.length === limit) {
                loggingProvider_1.logger.info(`... use the --limit flag to return a larger set`);
            }
            break;
    }
    return outputPlans.length;
}
function printPlansAsTable(plans) {
    const table = new cli_table3_1.default({
        head: ['ID', 'Name', 'Created time'],
        wordWrap: true,
    });
    plans.forEach((plan) => {
        table.push([
            { rowSpan: 1, content: plan.id, vAlign: 'center' },
            { rowSpan: 1, content: plan.name, vAlign: 'center' },
            {
                rowSpan: 1,
                content: moment.utc(plan.created_time).format(constants_1.ListTimeFormat),
                vAlign: 'center',
            },
        ]);
    });
    loggingProvider_1.logger.info(table.toString());
}
