// @mablhq/mabl-cli/reporters/__tests__/resources/sampleData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expectedPlainSuite = exports.sampleTestResult = void 0;
const execution_1 = require("../../../execution");
exports.sampleTestResult = {
    numFailedTests: 1,
    numPassedTests: 3,
    numSkippedTests: 1,
    numTotalTests: 5,
    startTime: 1627580108948,
    endTime: 1627580377201,
    totalTimeSeconds: 268.253,
    testResults: [
        {
            functionallyCompleted: false,
            status: execution_1.TestResultStatus.skipped,
            errorMessage: 'Test is not currently enabled',
            testId: 'jW60MgcXnf0qbw90AiXe1g-j:1',
            testName: 'Duplicate request headers',
            startTime: 1627580108956,
            endTime: 1627580108956,
        },
        {
            functionallyCompleted: true,
            status: execution_1.TestResultStatus.passed,
            testId: 'fb60BEqZa3kXZy1Gj0hAsA-j:4',
            testName: 'EX-001 - Click and Double Click',
            errorMessage: '',
            startTime: 1627580108962,
            endTime: 1627580114195,
        },
        {
            functionallyCompleted: false,
            status: execution_1.TestResultStatus.failed,
            testId: 'WvpAcSKyT2Sh5Rp817Saiw-j:0',
            testName: 'EX-002 - Assert presence',
            errorMessage: 'Failure running Flow:  - e4NjKPbkHut9ONLTFn75NA-f:1 - Element not found',
            startTime: 1627580114197,
            endTime: 1627580116478,
        },
        {
            functionallyCompleted: true,
            status: execution_1.TestResultStatus.passed,
            testId: 'PgRXOjMkUcqqLiRRLZVZMA-j:2',
            testName: 'EX-056 Enter text on HTML5 input types',
            errorMessage: '',
            startTime: 1627580369221,
            endTime: 1627580372959,
        },
        {
            functionallyCompleted: true,
            status: execution_1.TestResultStatus.passed,
            testId: 'gGxPiAi2XfJQuqSYc30xGA-j:0',
            testName: 'EX-058 Enter text into non-text input',
            errorMessage: '',
            startTime: 1627580372960,
            endTime: 1627580377201,
        },
    ],
    success: true,
};
exports.expectedPlainSuite = {
    _timeout: 0,
    afterHooks: [],
    beforeHooks: [],
    duration: 268253,
    failures: ['testUuid'],
    file: '',
    fullFile: '',
    passes: ['testUuid', 'testUuid', 'testUuid'],
    pending: [],
    root: true,
    rootEmpty: false,
    skipped: ['testUuid'],
    suites: [],
    tests: [
        {
            duration: 0,
            code: 'Test was skipped',
            err: { message: 'Test is not currently enabled' },
            fail: false,
            fullTitle: 'Duplicate request headers - jW60MgcXnf0qbw90AiXe1g-j:1',
            isHook: false,
            parentUUID: 'testUuid',
            pass: false,
            pending: false,
            skipped: true,
            state: undefined,
            timedOut: false,
            title: 'Duplicate request headers',
            uuid: 'testUuid',
        },
        {
            duration: 5233,
            code: 'Test passed',
            err: {},
            fail: false,
            fullTitle: 'EX-001 - Click and Double Click - fb60BEqZa3kXZy1Gj0hAsA-j:4',
            isHook: false,
            parentUUID: 'testUuid',
            pass: true,
            pending: false,
            skipped: false,
            state: 'passed',
            timedOut: false,
            title: 'EX-001 - Click and Double Click',
            uuid: 'testUuid',
        },
        {
            duration: 2281,
            code: 'Test has failed',
            err: {
                message: 'Failure running Flow:  - e4NjKPbkHut9ONLTFn75NA-f:1 - Element not found',
            },
            fail: true,
            fullTitle: 'EX-002 - Assert presence - WvpAcSKyT2Sh5Rp817Saiw-j:0',
            isHook: false,
            parentUUID: 'testUuid',
            pass: false,
            pending: false,
            skipped: false,
            state: 'failed',
            timedOut: false,
            title: 'EX-002 - Assert presence',
            uuid: 'testUuid',
        },
        {
            duration: 3738,
            code: 'Test passed',
            err: {},
            fail: false,
            fullTitle: 'EX-056 Enter text on HTML5 input types - PgRXOjMkUcqqLiRRLZVZMA-j:2',
            isHook: false,
            parentUUID: 'testUuid',
            pass: true,
            pending: false,
            skipped: false,
            state: 'passed',
            timedOut: false,
            title: 'EX-056 Enter text on HTML5 input types',
            uuid: 'testUuid',
        },
        {
            duration: 4241,
            code: 'Test passed',
            err: {},
            fail: false,
            fullTitle: 'EX-058 Enter text into non-text input - gGxPiAi2XfJQuqSYc30xGA-j:0',
            isHook: false,
            parentUUID: 'testUuid',
            pass: true,
            pending: false,
            skipped: false,
            state: 'passed',
            timedOut: false,
            title: 'EX-058 Enter text into non-text input',
            uuid: 'testUuid',
        },
    ],
    title: 'mabl run --id SPYP5dzZwu81x1xBwGL0Pg-j --reporter mochawesome --headless',
    uuid: 'testUuid',
};
