// @mablhq/mabl-cli/core/messaging/actions/pdfActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PdfEventEmitter = exports.PdfActionMessageType = void 0;
const messaging_1 = require("../messaging");
var PdfActionMessageType;
(function (PdfActionMessageType) {
    PdfActionMessageType["EmbeddedPdfDownloading"] = "EMBEDDED_PDF_DOWNLOADING";
    PdfActionMessageType["EmbeddedPdfViewerReady"] = "EMBEDDED_PDF_VIEWER_READY";
})(PdfActionMessageType || (exports.PdfActionMessageType = PdfActionMessageType = {}));
class PdfEventEmitter {
    static embeddedPdfDownloading(frameId, pdfFileName, pdfUrl) {
        PdfEventEmitter.eventEmitter.dispatch(new messaging_1.MablCoreAction(PdfActionMessageType.EmbeddedPdfDownloading, {
            frameId,
            pdfFileName,
            pdfUrl,
        }));
    }
    static embeddedPdfViewerReady(frameId, pdfFileName, pdfUrl) {
        PdfEventEmitter.eventEmitter.dispatch(new messaging_1.MablCoreAction(PdfActionMessageType.EmbeddedPdfViewerReady, {
            frameId,
            pdfFileName,
            pdfUrl,
        }));
    }
}
exports.PdfEventEmitter = PdfEventEmitter;
PdfEventEmitter.eventEmitter = messaging_1.mablEventEmitter;
