// @mablhq/mabl-cli/core/messaging/actions/trainingSessionActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrainingEventEmitter = exports.TrainingSessionMessageType = void 0;
const messaging_1 = require("../messaging");
var TrainingSessionMessageType;
(function (TrainingSessionMessageType) {
    TrainingSessionMessageType["closed"] = "TRAINING_SESSION_CLOSED";
    TrainingSessionMessageType["error"] = "TRAINING_SESSION_ERROR";
    TrainingSessionMessageType["initializing"] = "TRAINING_SESSION_INITIALIZING";
    TrainingSessionMessageType["saved"] = "TRAINING_SESSION_SAVED";
    TrainingSessionMessageType["started"] = "TRAINING_SESSION_STARTED";
})(TrainingSessionMessageType || (exports.TrainingSessionMessageType = TrainingSessionMessageType = {}));
class TrainingEventEmitter {
    static sessionInitializing() {
        messaging_1.mablEventEmitter.dispatch(new messaging_1.MablCoreAction(TrainingSessionMessageType.initializing, {}));
    }
    static sessionStarted() {
        messaging_1.mablEventEmitter.dispatch(new messaging_1.MablCoreAction(TrainingSessionMessageType.started, {}));
    }
    static sessionError(errorMessage) {
        messaging_1.mablEventEmitter.dispatch(new messaging_1.MablCoreAction(TrainingSessionMessageType.error, {
            errorMessage,
        }));
    }
    static sessionSaved(testId, testInvariantId, branchName, planId) {
        const actionPayload = {
            testId,
            testInvariantId,
            planId,
            branchName,
        };
        messaging_1.mablEventEmitter.dispatch(new messaging_1.MablCoreAction(TrainingSessionMessageType.saved, actionPayload));
    }
    static sessionClosed() {
        messaging_1.mablEventEmitter.dispatch(new messaging_1.MablCoreAction(TrainingSessionMessageType.closed, {}));
    }
}
exports.TrainingEventEmitter = TrainingEventEmitter;
