// @mablhq/mabl-cli/commands/mobile-build-files/mobile-build-files_cmds/delete.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mablApiClientFactory_1 = require("../../../api/mablApiClientFactory");
const constants_1 = require("../../constants");
const util_1 = require("../../commandUtil/util");
const chalk_1 = __importDefault(require("chalk"));
const loggingProvider_1 = require("../../../providers/logging/loggingProvider");
exports.command = `delete <${constants_1.CommandArgId}>`;
exports.describe = 'Delete a mobile build file.';
exports.builder = (yargs) => {
    yargs.positional(constants_1.CommandArgId, {
        describe: 'ID of the mobile build file',
        type: 'string',
    });
};
exports.handler = (0, util_1.failWrapper)(deleteMobileAppFile);
async function deleteMobileAppFile(parsed) {
    try {
        const apiClient = await mablApiClientFactory_1.MablApiClientFactory.createApiClient();
        const id = parsed[constants_1.CommandArgId];
        await apiClient.deleteMobileAppFile(id);
        loggingProvider_1.logger.info(chalk_1.default.green(`Mobile build file ${chalk_1.default.bold(id)} DELETED`));
        return id;
    }
    catch (error) {
        throw new Error(`Error deleting mobile build file: ${error}`);
    }
}
